junit-common/src/main/java/org/jboss/weld/junit/AbstractMockBean.java

```diff
@@ -0,0 +1,454 @@
+         * @return self
+         */
+        public K addStereotype(Class<? extends Annotation> stereotype) {
+            this.stereotypes.add(stereotype);
+            return self();
+        }
+
+        /**
+         * Each invocation of {@link Bean#create(CreationalContext)} will return the same instance.
+         *
+         * @param instance
+         * @return self
+         */
+        public K creating(final T instance) {
+            this.createCallback = ctx -> instance;
+            return self();
+        }
+
+        /**
+         * Use {@link Unmanaged} to create/destroy the bean instance.
+         *
+         * <p>
+         * NOTE: {@link CreationalContext#toString()} is used as a key in a map and therefore must be unique for the lifetime of a bean instance. Weld
+         * implementation fulfills this requirement.
+         * </p>
+         *
+         * @return self
+         * @see Unmanaged.UnmanagedInstance
+         */
+        @SuppressWarnings("unchecked")
+        public K useUnmanaged(Class<T> beanClass) {
+            Map<String, Unmanaged.UnmanagedInstance<?>> ctxToUnmanaged = new ConcurrentHashMap<>();
+            create(ctx -> {
+                Unmanaged<?> unmanaged = new Unmanaged<>(WeldContainer.current().getBeanManager(), beanClass);
+                Unmanaged.UnmanagedInstance<?> unmanagedInstance = unmanaged.newInstance();
+                ctxToUnmanaged.put(ctx.toString(), unmanagedInstance);
+                return (T) unmanagedInstance.produce().inject().postConstruct().get();
+            });
+            destroy((o, ctx) -> {
+                Unmanaged.UnmanagedInstance<?> unmanagedInstance = ctxToUnmanaged.remove(ctx.toString());
+                if (unmanagedInstance != null) {
+                    if (!unmanagedInstance.get().equals(o)) {
+                        throw new IllegalStateException(
+                                "Unmanaged instance [" + unmanagedInstance.get() + "] is not equal to the bean instance to be destroyed: " + o);
+                    }
+                    unmanagedInstance.preDestroy().dispose();
+                }
+            });
+            return self();
+        }
+
+        /**
+         *
+         * @param callback
+         * @return self
+         * @see Bean#create(CreationalContext)
+         */
+        public K create(CreateFunction<T> callback) {
+            this.createCallback = callback;
+            return self();
+        }
+
+        /**
+         *
+         * @param callback
+         * @return self
+         * @see Bean#destroy(Object, CreationalContext)
+         */
+        public K destroy(DestroyFunction<T> callback) {
+            this.destroyCallback = callback;
+            return self();
+        }
+
+        /**
+         *
+         * @return a new {@link AbstractMockBean} instance
+         * @throws IllegalStateException If a create callback is not set
+         */
+        public abstract AbstractMockBean<T> build();
```

Comment:
This should probably return `K`?

Reply:
Dismissing this as I reworked it completely.